fairscale/optim/oss.py

```diff
@@ -51,30 +53,37 @@ class OSS(Optimizer):
 
     in_super_constructor: bool
 
-    def __init__(self, params: _params_t, optim: Type[Optimizer] = SGD, group: Any = dist.group.WORLD, **defaults: Any):
+    def __init__(self, params: _params_t, optim: Type[Optimizer] = SGD, group: Optional[Any] = None, **default: Any):
         # Hold all the model params in the root .param_groups
         self.in_super_constructor = True
-        super().__init__(params, defaults)
+        super().__init__(params, default)
         self.in_super_constructor = False
 
+        # Partition information. lazy evaluation, computed if requested
+        self._per_device_params: List[List[Parameter]] = []
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[dict]] = []
+
         # Build the wrapped optimizer, responsible for a shard of the params
-        self.group = group
-        self.rank = dist.get_rank(group)
-        split_param_groups = self.partition_parameters()
-        self.optim = optim(split_param_groups[self.rank], **defaults)
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+
+        self.rank = dist.get_rank(self.group)
+        self.optim = optim(self.partition_parameters()[self.rank], **default)
 
         #  Optional consolidated optimizer state
         self._all_states: List[Dict[str, Any]] = []
 
-        #  Current device is set by the parameters allocated to this rank
-        self._device = split_param_groups[self.rank][0]["params"][0].device
+        # Current device is set by the parameters allocated to this rank
+        self._device = self.partition_parameters()[self.rank][0]["params"][0].device
 
         # Sync local and global param_groups keys
         for global_group, local_group in zip(self.param_groups, self.optim.param_groups):
             for k, v in local_group.items():
                 if k != "params":
                     global_group[k] = v
 
+    # Partition helpers
     def partition_parameters(self) -> List[List[dict]]:
```

Comment:
I think that was a little slow, because being re-run for all step(). Now it's cached and re-run only when we modify the param_groups

Reply:
could have been a property by the way, but I was not sure of the fairseq usecase, whether this was being accessed or not (cc @msbaines )